Logfiles laden

In [1]:
from src.loader import load_data

# Daten vorbereiten

In [19]:
import pm4py

# Load the logs
log_domestic = load_data('domestic')
log_international = load_data('international')
log_domestic_nonanomaly = pm4py.read_xes("Data/nonanomaly_DomesticDeclarations.xes")
log_international_nonanomaly = pm4py.read_xes("Data/nonanomaly_InternationalDeclarations.xes")

# Discover Petri net for Domestic - Normal
dom_petri_net, dom_initial_marking, dom_final_marking = pm4py.discover_petri_net_inductive(log_domestic)

# Discover Petri net for Domestic - Non-Anomaly
dom_petri_net_non_a, dom_initial_marking_non_a, dom_final_marking_non_a = pm4py.discover_petri_net_inductive(log_domestic_nonanomaly)

# Discover Petri net for International - Normal
int_petri_net, int_initial_marking, int_final_marking = pm4py.discover_petri_net_inductive(log_international)

# Discover Petri net for International - Non-Anomaly
int_petri_net_non_a, int_initial_marking_non_a, int_final_marking_non_a = pm4py.discover_petri_net_inductive(log_international_nonanomaly)

# Displaying the results for clarity
print("Petri net for Domestic - Normal:")
print(dom_petri_net)

print("\nPetri net for Domestic - Non-Anomaly:")
print(dom_petri_net_non_a)

print("\nPetri net for International - Normal:")
print(int_petri_net)

print("\nPetri net for International - Non-Anomaly:")
print(int_petri_net_non_a)


parsing log, completed traces ::   0%|          | 0/10500 [00:00<?, ?it/s]

parsing log, completed traces ::   0%|          | 0/6449 [00:00<?, ?it/s]

parsing log, completed traces ::   0%|          | 0/9899 [00:00<?, ?it/s]

parsing log, completed traces ::   0%|          | 0/3781 [00:00<?, ?it/s]

Petri net for Domestic - Normal:
places: [ p_10, p_11, p_12, p_13, p_14, p_15, p_17, p_18, p_19, p_20, p_21, p_22, p_23, p_24, p_26, p_27, p_28, p_29, p_3, p_30, p_4, p_5, p_6, p_7, sink, source ]
transitions: [ (04692a87-fa78-464d-8ad7-f4dc3769a25a, 'Payment Handled'), (0c8f6366-2117-41d4-a639-9006ac70d07b, 'Declaration FINAL_APPROVED by SUPERVISOR'), (131e0c1e-56de-4c45-a210-58b304a051df, 'Declaration FOR_APPROVAL by PRE_APPROVER'), (2afb5de9-32fc-4385-bfe2-65aec24d66c1, 'Request Payment'), (2c0844b7-a3d2-4e3c-a70d-d64610544f45, 'Declaration APPROVED by ADMINISTRATION'), (3ef7dbda-7108-43f0-b1a6-c0bf99bb2223, 'Declaration SUBMITTED by EMPLOYEE'), (463036fa-375c-479e-bedc-de3ca1e8671c, 'Declaration FOR_APPROVAL by ADMINISTRATION'), (49a70374-c47d-4418-be19-44b5a956ba4d, 'Declaration REJECTED by SUPERVISOR'), (6832fc8f-639c-4689-87ca-84b9b0bcbb14, 'Declaration REJECTED by PRE_APPROVER'), (6eb91a53-29b1-4030-a1ef-b5fc45b9928b, 'Declaration REJECTED by BUDGET OWNER'), (84d825a0-5ed7-436e

# Process Conformance

## Fitness

### Token-based Replay Fitness

In [21]:
import pm4py

# Token-based Replay Fitness for Domestic - Normal
fitness_domestic_normal = pm4py.fitness_token_based_replay(
    log_domestic, 
    dom_petri_net, 
    dom_initial_marking, 
    dom_final_marking
)

# Token-based Replay Fitness for Domestic - Non-Anomaly
fitness_domestic_nonanomaly = pm4py.fitness_token_based_replay(
    log_domestic_nonanomaly, 
    dom_petri_net_non_a, 
    dom_initial_marking_non_a, 
    dom_final_marking_non_a
)

# Token-based Replay Fitness for International - Normal
fitness_international_normal = pm4py.fitness_token_based_replay(
    log_international, 
    int_petri_net, 
    int_initial_marking, 
    int_final_marking
)

# Token-based Replay Fitness for International - Non-Anomaly
fitness_international_nonanomaly = pm4py.fitness_token_based_replay(
    log_international_nonanomaly, 
    int_petri_net_non_a, 
    int_initial_marking_non_a, 
    int_final_marking_non_a
)

# Displaying the fitness results
print("Token-based Replay Fitness for Domestic - Normal:")
print(fitness_domestic_normal)

print("\nToken-based Replay Fitness for Domestic - Non-Anomaly:")
print(fitness_domestic_nonanomaly)

print("\nToken-based Replay Fitness for International - Normal:")
print(fitness_international_normal)

print("\nToken-based Replay Fitness for International - Non-Anomaly:")
print(fitness_international_nonanomaly)


replaying log with TBR, completed traces ::   0%|          | 0/99 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/8 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/753 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/13 [00:00<?, ?it/s]

Token-based Replay Fitness for Domestic - Normal:
{'perc_fit_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 1.0, 'percentage_of_fitting_traces': 100.0}

Token-based Replay Fitness for Domestic - Non-Anomaly:
{'perc_fit_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 1.0, 'percentage_of_fitting_traces': 100.0}

Token-based Replay Fitness for International - Normal:
{'perc_fit_traces': 6.946813459451078, 'average_trace_fitness': 0.971365449043232, 'log_fitness': 0.9707533277618512, 'percentage_of_fitting_traces': 6.946813459451078}

Token-based Replay Fitness for International - Non-Anomaly:
{'perc_fit_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 1.0, 'percentage_of_fitting_traces': 100.0}


- **Domestic - Normal**: 
  - Alle Prozessvarianten passen perfekt zum Modell. Der Fitness-Wert von 1.0 für alle Metriken zeigt, dass das Modell den realen Prozess vollständig repräsentiert. Es gibt keine Anomalien oder Abweichungen in den Daten.

- **Domestic - Non-Anomaly**:
  - Auch nach dem Entfernen der seltenen Varianten (weniger als 1% der Gesamttraces) bleibt die Fitness perfekt. Dies zeigt, dass die Anomalien keinen Einfluss auf die allgemeine Prozessstruktur hatten.

- **International - Normal**:
  - Nur etwa 6.95% der Traces passen vollständig zum Modell. Der durchschnittliche Trace-Fitness-Wert von 0.971 zeigt jedoch, dass die meisten Traces dem Modell recht nahe kommen, aber es gibt signifikante Abweichungen oder Varianten, die nicht gut durch das Modell erfasst werden.

- **International - Non-Anomaly**:
  - Nach dem Entfernen der seltenen Varianten passen alle verbleibenden Traces perfekt zum Modell. Dies deutet darauf hin, dass die Anomalien hauptsächlich für die Abweichungen im normalen Datensatz verantwortlich waren.

### Alignment-based Fitness

In [23]:
import pm4py

# Alignment-based Fitness for Domestic - Normal
alignment_domestic_normal = pm4py.fitness_alignments(
    log_domestic, 
    dom_petri_net, 
    dom_initial_marking, 
    dom_final_marking,
    multi_processing=True
)

# Alignment-based Fitness for Domestic - Non-Anomaly
alignment_domestic_nonanomaly = pm4py.fitness_alignments(
    log_domestic_nonanomaly, 
    dom_petri_net_non_a, 
    dom_initial_marking_non_a, 
    dom_final_marking_non_a,
    multi_processing=True
)

# Alignment-based Fitness for International - Normal
alignment_international_normal = pm4py.fitness_alignments(
    log_international, 
    int_petri_net, 
    int_initial_marking, 
    int_final_marking,
    multi_processing=True
)

# Alignment-based Fitness for International - Non-Anomaly
alignment_international_nonanomaly = pm4py.fitness_alignments(
    log_international_nonanomaly, 
    int_petri_net_non_a, 
    int_initial_marking_non_a, 
    int_final_marking_non_a,
    multi_processing=True
)

# Displaying the fitness results
print("Alignment-based Fitness for Domestic - Normal:")
print(alignment_domestic_normal)

print("\nAlignment-based Fitness for Domestic - Non-Anomaly:")
print(alignment_domestic_nonanomaly)

print("\nAlignment-based Fitness for International - Normal:")
print(alignment_international_normal)

print("\nAlignment-based Fitness for International - Non-Anomaly:")
print(alignment_international_nonanomaly)


aligning log, completed variants ::   0%|          | 0/99 [00:00<?, ?it/s]

aligning log, completed variants ::   0%|          | 0/8 [00:00<?, ?it/s]

aligning log, completed variants ::   0%|          | 0/753 [00:00<?, ?it/s]

aligning log, completed variants ::   0%|          | 0/13 [00:00<?, ?it/s]

Alignment-based Fitness for Domestic - Normal:
{'percFitTraces': 100.0, 'averageFitness': 1.0, 'percentage_of_fitting_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 0.9997527210745741}

Alignment-based Fitness for Domestic - Non-Anomaly:
{'percFitTraces': 100.0, 'averageFitness': 1.0, 'percentage_of_fitting_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 0.9999096643560642}

Alignment-based Fitness for International - Normal:
{'percFitTraces': 100.0, 'averageFitness': 1.0, 'percentage_of_fitting_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 0.9996963994751248}

Alignment-based Fitness for International - Non-Anomaly:
{'percFitTraces': 100.0, 'averageFitness': 1.0, 'percentage_of_fitting_traces': 100.0, 'average_trace_fitness': 1.0, 'log_fitness': 0.9999430903417306}


- **Domestic - Normal**:
  - Das Modell repräsentiert den realen Prozess fast perfekt, wobei alle Traces vollständig zum Modell passen. Der Log-Fitness-Wert von 0.99975 zeigt, dass es nur minimale Abweichungen gibt, die jedoch vernachlässigbar sind.

- **Domestic - Non-Anomaly**:
  - Nach dem Entfernen der seltenen Varianten zeigt das Modell eine nahezu perfekte Übereinstimmung mit den verbleibenden Daten, was durch den leicht verbesserten Log-Fitness-Wert von 0.99991 bestätigt wird.

- **International - Normal**:
  - Trotz der Komplexität des internationalen Prozesses zeigt das Modell eine fast perfekte Übereinstimmung mit den Traces. Der Log-Fitness-Wert von 0.99969 deutet darauf hin, dass es nur sehr geringe Abweichungen gibt.

- **International - Non-Anomaly**:
  - Durch das Entfernen der seltenen Varianten wird eine nahezu perfekte Fitness von 0.99994 erreicht. Dies unterstreicht, dass die Anomalien in den normalen Daten für die geringen Abweichungen verantwortlich waren.

Zusammenfassend zeigen die Fitness-Analysen, sowohl auf Token- als auch auf Alignment-Basis, dass die Modelle die Prozesse insgesamt sehr gut repräsentieren. Insbesondere nach dem Entfernen der seltenen Anomalien erreichen alle Modelle nahezu perfekte Fitness-Werte, was darauf hindeutet, dass die Hauptprozesse konsistent und korrekt erfasst wurden. Die geringen Abweichungen in den Fitness-Werten der normalen Datensätze sind größtenteils auf seltene Prozessvarianten zurückzuführen, die als Anomalien identifiziert wurden.

## Precision Token-based Replay

In [24]:
import pm4py

# Precision Token-based Replay für Domestic - Normal
precision_domestic_normal = pm4py.conformance.precision_token_based_replay(
    log_domestic, dom_petri_net, dom_initial_marking, dom_final_marking
)

# Precision Token-based Replay für Domestic - Non-Anomaly
precision_domestic_nonanomaly = pm4py.conformance.precision_token_based_replay(
    log_domestic_nonanomaly, dom_petri_net_non_a, dom_initial_marking_non_a, dom_final_marking_non_a
)

# Precision Token-based Replay für International - Normal
precision_international_normal = pm4py.conformance.precision_token_based_replay(
    log_international, int_petri_net, int_initial_marking, int_final_marking
)

# Precision Token-based Replay für International - Non-Anomaly
precision_international_nonanomaly = pm4py.conformance.precision_token_based_replay(
    log_international_nonanomaly, int_petri_net_non_a, int_initial_marking_non_a, int_final_marking_non_a
)

# Ergebnisse anzeigen
print("Precision Token-based Replay für Domestic - Normal:")
print(precision_domestic_normal)

print("\nPrecision Token-based Replay für Domestic - Non-Anomaly:")
print(precision_domestic_nonanomaly)

print("\nPrecision Token-based Replay für International - Normal:")
print(precision_international_normal)

print("\nPrecision Token-based Replay für International - Non-Anomaly:")
print(precision_international_nonanomaly)


replaying log with TBR, completed traces ::   0%|          | 0/324 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/21 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/4077 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/89 [00:00<?, ?it/s]

Precision Token-based Replay für Domestic - Normal:
0.3284064804374959

Precision Token-based Replay für Domestic - Non-Anomaly:
0.7002038553041494

Precision Token-based Replay für International - Normal:
0.1593363814500327

Precision Token-based Replay für International - Non-Anomaly:
0.35798937642537076


Die Precision-Ergebnisse des Token-based Replay zeigen, dass die Modelle bei den **Domestic-Daten** deutlich präziser sind als bei den **International-Daten**. Insbesondere nach dem Entfernen der seltenen Varianten (**Non-Anomaly**) steigt die Precision bei beiden Prozessen, was darauf hinweist, dass die Anomalien die Präzision des Modells signifikant beeinträchtigt haben. Allerdings bleibt die Precision insgesamt relativ niedrig, was darauf hindeutet, dass die Modelle noch unnötiges oder überflüssiges Verhalten zulassen.

## Generalization

In [29]:
import pm4py

# Generalization Token-based Replay für Domestic - Normal
generalization_domestic_normal = pm4py.conformance.generalization_tbr(
    log_domestic, dom_petri_net, dom_initial_marking, dom_final_marking
)

# Generalization Token-based Replay für Domestic - Non-Anomaly
generalization_domestic_nonanomaly = pm4py.conformance.generalization_tbr(
    log_domestic_nonanomaly, dom_petri_net_non_a, dom_initial_marking_non_a, dom_final_marking_non_a
)

# Generalization Token-based Replay für International - Normal
generalization_international_normal = pm4py.conformance.generalization_tbr(
    log_international, int_petri_net, int_initial_marking, int_final_marking
)

# Generalization Token-based Replay für International - Non-Anomaly
generalization_international_nonanomaly = pm4py.conformance.generalization_tbr(
    log_international_nonanomaly, int_petri_net_non_a, int_initial_marking_non_a, int_final_marking_non_a
)

# Ergebnisse anzeigen
print("Generalization Token-based Replay für Domestic - Normal:")
print(generalization_domestic_normal)

print("\nGeneralization Token-based Replay für Domestic - Non-Anomaly:")
print(generalization_domestic_nonanomaly)

print("\nGeneralization Token-based Replay für International - Normal:")
print(generalization_international_normal)

print("\nGeneralization Token-based Replay für International - Non-Anomaly:")
print(generalization_international_nonanomaly)


replaying log with TBR, completed traces ::   0%|          | 0/99 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/8 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/753 [00:00<?, ?it/s]

replaying log with TBR, completed traces ::   0%|          | 0/13 [00:00<?, ?it/s]

Generalization Token-based Replay für Domestic - Normal:
0.8554073029005208

Generalization Token-based Replay für Domestic - Non-Anomaly:
0.9744419345910126

Generalization Token-based Replay für International - Normal:
0.8766720880937127

Generalization Token-based Replay für International - Non-Anomaly:
0.9661188835125442


Die Generalization-Ergebnisse des Token-based Replay zeigen, dass die Modelle eine gute Generalisierungsfähigkeit besitzen, insbesondere nach dem Entfernen der seltenen Varianten (Nicht-Anomaly). Sowohl bei den Domestic- als auch bei den International-Daten verbessern sich die Generalisierungswerte deutlich, was darauf hindeutet, dass die Modelle nach dem Ausschluss von Anomalien besser in der Lage sind, plausible zukünftige Prozessvarianten abzudecken. Die Generalization-Werte liegen insgesamt auf einem hohen Niveau, was auf eine robuste Modellierung der Prozesse hinweist.